### PR TITLE
Create a component migration script

### PR DESCRIPTION
**Pull request by bkjohnson**
## Description

This is a migration script + transformer for `<AlertBox>` => `<va-alert>`. The overall script is written in a way so that we can add new transformers for each component that we want to migrate.

## Testing done

Local testing under the following conditions:

1. Single component in a file
1. multiple components in a file
1. "shortened" React components, i.e. `<AlertBox prop1="foo" />` as opposed to `<AlertBox prop1="foo">content</AlertBox>`
1. React components with a named closing tag
1. simple prop renaming
1. prop remapping using a function

## Screenshots

A sample of the migration for `AlertBox`:

```bash
yarn migrate-component --dir script/component-migration/samples/ --component AlertBox
```

image

## Acceptance criteria
- [ ]

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: script/component-migration/index.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+function replaceTags(file, newTag) {
+  const unnamedClosingTags = file.matchAll(
+    new RegExp(`(<${options.component}.+?(^\\s+)?\\/>;?$)`, 'gsm'),
+  );
+
+  const matches = Array.from(unnamedClosingTags, m => m[0]);
+
+  let namedClosingTags = file;
+  matches.forEach(match => {
```
Comment (cvalarida) [inline script/component-migration/index.js:60]: You can accomplish the same thing with `matches.reduce()`.

File: script/component-migration/transformers/alertbox.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
+function alertBoxReplacement() {
+  const moveToNamedSlot = componentString => {
+    let translatedHeadline = componentString.toString();
+    const level = componentString.match(/level=["{](\d)["}]/)?.[1] || 3;
+    /* eslint-disable-next-line no-useless-escape */
```
[the remark this answers is not in the document]
Comment (cvalarida) [reply script/component-migration/transformers/alertbox.js:5]: Do you need this?

**Review by cvalarida — COMMENTED** [review]
There's some cleanup you could do here, but I think this is a good first pass. It does the job pretty well, from the tests I've run. :+1: There are a few edge cases I ran into, but those can be fixed manually pretty easily.

I'd like to see two more things before I approve:
1. External documentation. Instructions on how to run the script, a word of warning on verifying the output since there are likely to be edge cases, a `--help` for the script, and meaningful error messages where appropriate. Because we want folks to run this themselves, we've gotta make sure the interface for using the tool is polished. First impressions can be hard to change, so I want to make sure that part is solid. :sweat_smile:
2. Internal documentation. A brief JSDoc comment on each function explaining its purpose and maybe execution, and then a short README for how to write new transformers for other components. I recommend focusing on the inputs and the outputs for that kind of documentation.

If you'd like to add those :point_up: in a follow-up PR, that's also okay with me. Let me know and I can approve this and await the follow-up.

File: script/component-migration/index.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  options.component
+}'`;
+
+function handleError(error) {
+  console.log(error);
+}
+
+function translateProps(componentString, propMap) {
+  // Make a copy so we don't modify the reference
+  let translatedComp = componentString.toString();
```
[the remark this answers is not in the document]
Comment (bkjohnson) [reply script/component-migration/index.js:28]: I was just playing it safe, but I'll give it a shot without the `.toString()`

File: script/component-migration/index.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+      switch (typeof newValue) {
+        case 'string':
+          translatedComp = translatedComp.replace(prop, newValue);
+          break;
+        case 'function':
+          translatedComp = newValue(translatedComp, prop);
+          break;
+        default:
+          break;
+      }
```
[the remark this answers is not in the document]
Comment (bkjohnson) [reply script/component-migration/index.js:46]: I went with a switch thinking that we might have all of these, and also a `null` case, possibly more. But simpler is better for now.

File: script/component-migration/transformers/alertbox.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (bkjohnson) [reply script/component-migration/transformers/alertbox.js:5]: Not anymore apparently!

File: script/component-migration/index.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+function translateProps(componentString, propMap) {
+  // Make a copy so we don't modify the reference
+  let translatedComp = componentString.toString();
+
+  Object.entries(propMap)
+    .sort((pairOne, _) => {
+      const children = ['content', 'children'];
+      if (children.includes(pairOne[0])) return -1;
```
[the remark this answers is not in the document]
Comment (bkjohnson) [reply script/component-migration/index.js:32]: `pairOne` refers to the first item to be comparred in the comparator function: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#parameters

In this case it's a pair because `Object.entries` gives us a pair: the key, and the value. Would `propPair` be more helpful?

File: script/component-migration/index.js [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+          break;
+      }
+    });
+
+  return translatedComp;
+}
+
+function replaceTags(file, newTag) {
+  const unnamedClosingTags = file.matchAll(
+    new RegExp(`(<${options.component}.+?(^\\s+)?\\/>;?$)`, 'gsm'),
```
[the remark this answers is not in the document]
Comment (bkjohnson) [reply script/component-migration/index.js:47]: This optionally targets components where the unnamed closing tag (` />`) is on a newline as opposed to being at the end of a short line.

File: script/component-migration/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (bkjohnson) [reply script/component-migration/index.js]: I went ahead and destrctured the pair so we can work with `prop` directly :shrug:

File: script/component-migration/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (cvalarida) [reply script/component-migration/index.js:63]: Ohhhhhhh, yeah yeah. I see it now. Having `^` in the middle of a RegEx really threw me for a loop. :sweat_smile: I don't think it's strictly-speaking needed, but that's fine.

File: script/component-migration/README.md [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+- `--dir` - The directory you wish to migrate
+- `--component` - The name of the component you wish to migrate
+
+If these aren't provided or `--help` is passed, the script will print out some help text describing how to use the arguments.
+
+Once the script finishes a `post-` hook will run to lint any changed files. The migration may have introduced unwanted whitespace, so this will clean that up and (hopefully) get your files ready to be committed.
+
+## Caveats
+
+The script may not be able to perfectly migrate every comopnent, so before you merge your PR be sure to inspect the diff. Additionally the script does not migrate test files - you will have to do those by hand.
```
Comment (cvalarida) [inline script/component-migration/README.md:14]:
```suggestion
**The script may not be able to perfectly migrate every component.** Before you merge your PR be sure to inspect the diff. Additionally the script does not migrate test files, so you'll have to do those by hand.
```

File: script/component-migration/README.md [hunk 8]
```diff
@@ -0,0 +1,14 @@
+## Component migration script
+
+The name of the script is `migrate-component`. It has two required arguments:
```
Comment (cvalarida) [inline script/component-migration/README.md:3]:
```suggestion
The purpose of this script is to do the bulk of the grunt work of migrating from an existing React component in the `@department-of-veterans-affairs/component-library` to its replacement web component.

To run the script, use `yarn migrate-component`. It accepts two required arguments:
```

File: script/component-migration/index.js [hunk 9]
```diff
@@ -23,52 +23,66 @@ function handleError(error) {
...
+  let translatedComp = componentString;
 
+  // We want any `children` or `content` props to be the first to migrate,
+  // because not doing so can mess up the closing tags
   Object.entries(propMap)
-    .sort((pairOne, _) => {
+    .sort(([prop, _value]) => {
       const children = ['content', 'children'];
-      if (children.includes(pairOne[0])) return -1;
+      if (children.includes(prop)) return -1;
```
Comment (cvalarida) [inline script/component-migration/index.js]: Ah, this is much cleaner. Thanks! :raised_hands:

File: script/component-migration/index.js [hunk 10]
```diff
@@ -23,52 +23,66 @@ function handleError(error) {
...
 
   return translatedComp;
 }
 
-function replaceTags(file, newTag) {
-  const unnamedClosingTags = file.matchAll(
+/**
+ * Replace the React component tags with the Web Component tag.
+ * This works on single line components, multiline components, and
+ * ones that don'e have a closing tag as well as ones that do.
```
Comment (cvalarida) [inline script/component-migration/index.js:56]:
```suggestion
 * ones that don't have a closing tag as well as ones that do.
```

File: script/component-migration/index.js [hunk 11]
```diff
@@ -105,13 +119,26 @@ function migrateFile(fname, data) {
...
 
-FILENAMES.forEach(fname => {
-  fs.readFile(fname, 'utf8', (err, data) => {
-    if (err) {
-      return handleError(err);
-    }
-    if (!data.includes(legacyImport)) return null;
-    // Leave this file alone if it doesn't import the component
-    return migrateFile(fname, data);
+function main() {
```
Comment (cvalarida) [inline script/component-migration/index.js:122]: Alternatively, you could use an IIFE, but this is totally fine. :+1:

**Review by cvalarida — APPROVED** [review]
Looks great! :shipit:

File: script/component-migration/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (bkjohnson) [reply script/component-migration/index.js:63]: It _is_ a bit odd how it's contextual. Maybe it's a `not`, or maybe it means "start" :upside_down_face:

**Review by k80bowman — APPROVED** [review]
